### PR TITLE
add client call

### PR DESCRIPTION
**Pull request by BusyJay**
This pr adds functions needed to handle rpc on server side.

### CODE REVIEW

File: src/call/client.rs [hunk 1]
```diff
@@ -0,0 +1,346 @@
...
+use protobuf::{self, Message, MessageStatic};
+
+use async::{BatchMessage, CqFuture};
+use call::Call;
+use error::Error;
+use super::{SinkBase, StreamingBase};
+
+/// Update the flag bit in res.
+#[inline]
+fn change_flag(res: &mut u32, flag: u32, set: bool) {
```
Comment (andelf) [inline src/call/client.rs]: better use https://crates.io/crates/bitflags

File: src/call/client.rs [hunk 2]
```diff
@@ -0,0 +1,346 @@
...
+        self
+    }
+
+    /// Get the timeout.
+    pub fn timeout(&self) -> Option<Duration> {
+        self.timeout
+    }
+}
+
+/// A handler to handle uanry async call.
```
Comment (andelf) [inline src/call/client.rs]: typo

File: src/call/client.rs [hunk 3]
```diff
@@ -0,0 +1,346 @@
...
+#[derive(Default)]
+pub struct CallOption {
+    timeout: Option<Duration>,
+    write_flags: u32,
+    call_flags: u32,
+}
+
+impl CallOption {
+    /// Signal that the call is idempotent
+    pub fn with_idempotent(mut self, is_idempotent: bool) -> CallOption {
```
Comment (andelf) [inline src/call/client.rs]: `with_` means constructor, use `idempotent` as a builder chaining function name is ok

File: src/call/client.rs [hunk 4]
```diff
@@ -0,0 +1,346 @@
...
+    }
+}
+
+/// A handler to handle uanry async call.
+///
+/// The future is resolved once response is received.
+pub struct UnaryCallHandler<T> {
+    call: Call,
+    resp_f: CqFuture<BatchMessage>,
+    _resp: PhantomData<T>,
```
Comment (andelf) [inline src/call/client.rs]: it's always named `_marker`.

File: src/call/client.rs [hunk 5]
```diff
@@ -0,0 +1,346 @@
...
+    fn poll(&mut self) -> Poll<T, Error> {
+        let data = try_ready!(self.resp_f.poll());
+        let t = try!(protobuf::parse_from_bytes(&data.unwrap()));
+        Ok(Async::Ready(t))
+    }
+}
+
+/// A handler for client streaming call.
+///
+/// Once all request is flushed, it can be converted to `UnaryResponseReceiver`
```
Comment (andelf) [inline src/call/client.rs]: requests

File: src/call/client.rs [hunk 6]
```diff
@@ -0,0 +1,346 @@
...
+/// to receive response asynchronously.
+pub struct ClientStreamingCallHandler<P, Q> {
+    call: Call,
+    resp_f: CqFuture<BatchMessage>,
+    sink_base: SinkBase,
+    _req: PhantomData<P>,
+    _resp: PhantomData<Q>,
+}
```
Comment (andelf) [inline src/call/client.rs]: for type-safty, use PhantomData<(P,Q)> or sth.

File: src/call/client.rs (hunk 4 above)
Comment (BusyJay) [inline src/call/client.rs]: I don't think it's true. Name it `_resp` here to indicate that T is the type of response.

File: src/call/client.rs (hunk 6 above)
Comment (BusyJay) [inline src/call/client.rs]: Why is it related to type-safety?

File: src/call/client.rs (hunk 1 above)
Comment (BusyJay) [inline src/call/client.rs]: I think it's fine here. It's only for internal usage.

Comment (siddontang) [inline src/call/client.rs]: should add a test for this individual function.

File: src/call/client.rs [hunk 7]
```diff
@@ -0,0 +1,346 @@
...
+    type Error = Error;
+
+    fn poll(&mut self) -> Poll<T, Error> {
+        let data = try_ready!(self.resp_f.poll());
+        let t = try!(protobuf::parse_from_bytes(&data.unwrap()));
+        Ok(Async::Ready(t))
+    }
+}
+
+/// An unary response receiver. It's used for client streaming request.
```
Comment (disksing) [inline src/call/client.rs]: A unary

**Comment by disksing** [comment]
LGTM

File: src/call/client.rs (hunk 6 above)
Comment (andelf) [inline src/call/client.rs]: Forget about it.

**Pull request merged**